visual-java/src/main/java/com/saucelabs/visual/VisualApi.java

```diff
@@ -729,6 +752,54 @@ private DiffingMethodTolerance getDiffingMethodTolerance(CheckOptions checkOptio
     return sensitivity != null ? sensitivity : this.diffingMethodTolerance;
   }
 
+  private WebElement getClipElement(CheckOptions checkOptions) {
+    if (checkOptions.getClipElement() != null) {
+      return checkOptions.getClipElement();
+    } else if (checkOptions.getClipSelector() != null) {
+      return this.driver.findElement(By.cssSelector(checkOptions.getClipSelector()));
+    }
+
+    return null;
+  }
+
+  private byte[] cropImage(
+      byte[] imageData, String imageFormat, int x, int y, int width, int height) {
+    // We can check that before loading the image
+    if (x < 0 || y < 0) {
+      throw new VisualApiException("Clipped element is not visible.");
+    }
```

Comment:
Should we instead set `x` and/or `y` to 0, and clip what we see?
Same question for the next check about `imageWidth` and `imageHeight`.

Note that this can result in an empty image with dimensions of 0x0.

Reply:
In our other SDKs and the backend we don't throw if the clip element cannot be found, we just silently continue. It's also possible to have a `x` and `y` less than zero but the element still be visible on the page, no? Such as being slightly off screen? Though, it also looks like we scroll the page to the element prior to clipping when not in FPS to get it as much in screen as possible. May want to do that here prior to getting your scroll offsets.

I don't think we're preemptively handling any cases like this in the other SDKs. If it crops to be really small, well, that is what was requested.

Reply:
I think that's fine. IIRC selenium throws an exception with `findElement` / uses it as an assertion so that would throw if it were an invalid element anyway. I think that's fine if it would result in an invalid crop